connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/SparkSession.scala

```diff
@@ -331,14 +331,41 @@ class SparkSession private[sql] (
     newDataset(UnboundRowEncoder)(f)
   }
 
+  @Since("4.0.0")
+  @DeveloperApi
+  def newDataFrame(references: Seq[Column])(f: proto.Relation.Builder => Unit): DataFrame = {
```

Comment:
We should add a bit of documentation explaining why this is needed. It is not an entirely trivial API :)...